File: ios/Classes/FlutterAppsflyerPlugin.m
#import "FlutterAppsflyerPlugin.h"

@implementation FlutterAppsflyerPlugin

NSString* DONE = @"DONE";

static FlutterMethodChannel* afChannel;

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  afChannel = [FlutterMethodChannel
               methodChannelWithName:@"com.sarpongkb/flutter_appsflyer"
               binaryMessenger:[registrar messenger]];
  FlutterAppsflyerPlugin* instance = [[FlutterAppsflyerPlugin alloc] init];
  [registrar addApplicationDelegate:instance];
  [registrar addMethodCallDelegate:instance channel:afChannel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"initSdk" isEqualToString:call.method]) {
    [self onInitSdk:call result:result];
  } else if ([@"trackEvent" isEqualToString:call.method]) {
    [self onTrackEvent:call result:result];
  } else if ([@"getAppsflyerDeviceId" isEqualToString:call.method]) {
    [self onGetAppsflyerDeviceId:call result:result];
  } else if ([@"getSdkVersion" isEqualToString:call.method]) {
    [self onGetSdkVersion:call result:result];
  } else if ([@"generateShareLink" isEqualToString:call.method]) {
    [self onGenerateShareLink:call result:result];
  } else if ([@"setCustomerUserId" isEqualToString:call.method]) {
    [self onSetCustomerUserId:call result:result];
  } else if ([@"setUserEmails" isEqualToString:call.method]) {
    [self onSetUserEmails:call result:result];
  } else if ([@"setAdditionalData" isEqualToString:call.method]) {
    [self onSetAdditionalData:call result:result];
  } else if ([@"isDeviceTrackingEnabled" isEqualToString:call.method]) {
    [self isDeviceTrackingEnabled:call result:result];
  } else if ([@"setDeviceTrackingDisabled" isEqualToString:call.method]) {
    [self onSetDeviceTrackingDisabled:call result:result];
  } else {
    result(FlutterMethodNotImplemented);
  }
}


- (void) onInitSdk:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString* devKey = call.arguments[@"devKey"];
  NSString* appleAppId = call.arguments[@"appleAppId"];
  [AppsFlyerTracker sharedTracker].appsFlyerDevKey = devKey;
  [AppsFlyerTracker sharedTracker].appleAppID = appleAppId;
  [AppsFlyerTracker sharedTracker].delegate = self;
  [AppsFlyerTracker sharedTracker].isDebug = true;
  result(DONE);
}


- (void) onTrackEvent:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString* eventName = call.arguments[@"eventName"];
  NSDictionary* eventValues = call.arguments[@"eventValues"];
  [[AppsFlyerTracker sharedTracker] trackEvent:eventName withValues:eventValues];
  result(DONE);
}


- (void) onGetAppsflyerDeviceId:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString* deviceId = [[AppsFlyerTracker sharedTracker] getAppsFlyerUID];
  result(deviceId);
}


- (void) onGetSdkVersion:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString* sdkVersion = [[AppsFlyerTracker sharedTracker] getSDKVersion];
  result(sdkVersion);
}


- (void) isDeviceTrackingEnabled:(FlutterMethodCall*)call result:(FlutterResult)result {
  BOOL enabled = ![AppsFlyerTracker sharedTracker].deviceTrackingDisabled;
  result([NSNumber numberWithBool:enabled]);
}


- (void) onSetDeviceTrackingDisabled:(FlutterMethodCall*)call result:(FlutterResult)result {
  BOOL disabled = call.arguments;
  NSLog(@"Requesting Appsflyer deviceTrackingDisabled: %i", disabled);
  NSLog(@"Before: Appsflyer deviceTrackingDisabled: %i", [AppsFlyerTracker sharedTracker].deviceTrackingDisabled);
  [AppsFlyerTracker sharedTracker].deviceTrackingDisabled = disabled;
  NSLog(@"After: Appsflyer deviceTrackingDisabled: %i", [AppsFlyerTracker sharedTracker].deviceTrackingDisabled);
  result(DONE);
}

- (void) onGenerateShareLink:(FlutterMethodCall*)call result:(FlutterResult)result {
  NSString* campaign = call.arguments[@"campaign"];
  NSString* channel = call.arguments[@"channel"];
  NSString* playerId = call.arguments[@"playerId"];
  NSString* keywords = call.arguments[@"keywords"];

  [AppsFlyerShareInviteHelper generateInviteUrlWithLinkGenerator:^AppsFlyerLinkGenerator * _Nonnull(AppsFlyerLinkGenerator * _Nonnull generator) {
    [generator setCampaign:campaign];
    [generator setChannel:channel];
    [generator setReferrerCustomerId:playerId];
    [generator addParameterValue:keywords forKey:@"af_keywords"];
    return generator;
  } completionHandler:^(NSURL * _Nullable url) {
    result([url absoluteString]);
  }];
}


- (void) onSetCustomerUserId:(FlutterMethodCall*)call result:(FlutterResult)result {
  [[AppsFlyerTracker sharedTracker] setCustomerUserID:call.arguments];
  result(DONE);
}


- (void) onSetUserEmails:(FlutterMethodCall*)call result:(FlutterResult)result {
  [[AppsFlyerTracker sharedTracker] setUserEmails:call.arguments withCryptType:EmailCryptTypeNone];
  result(DONE);
}


- (void) onSetAdditionalData:(FlutterMethodCall*)call result:(FlutterResult)result {
  [[AppsFlyerTracker sharedTracker] setAdditionalData:call.arguments];
  result(DONE);
}


- (void) onConversionDataReceived:(NSDictionary *)installData {
  NSLog(@"FlutterAppsflyer onConversionDataReceived %@", installData);
  [afChannel invokeMethod:@"conversionDataReceived" arguments:installData];
}

- (void) onAppOpenAttribution:(NSDictionary *)attributionData {
  NSLog(@"FlutterAppsflyer onAppOpenAttribution %@", attributionData);
  [afChannel invokeMethod:@"appOpenAttribution" arguments:attributionData];
}

- (void)onAppOpenAttributionFailure:(NSError *)error {
  NSLog(@"FlutterAppsflyer onAppOpenAttributionFailure %li: %@", error.code, error.localizedDescription);
  [afChannel invokeMethod:@"appOpenAttributionFailure" arguments:error.localizedDescription];
}

- (void)onConversionDataRequestFailure:(NSError *)error {
  NSLog(@"FlutterAppsflyer onConversionDataRequestFailure %li: %@", error.code, error.localizedDescription);
  [afChannel invokeMethod:@"conversionDataRequestFailure" arguments:error.localizedDescription];
}


// Reports app open from deep link for iOS 10
- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url options:(NSDictionary *) options {
  [[AppsFlyerTracker sharedTracker] handleOpenUrl:url options:options];
  return YES;
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
  [[AppsFlyerTracker sharedTracker] trackAppLaunch];
}


// Reports app open from a Universal Link for iOS 9 or above
- (BOOL)application:(UIApplication *)application continueUserActivity:(NSUserActivity *)userActivity restorationHandler:(void (^)(NSArray * _Nonnull))restorationHandler
{
  [[AppsFlyerTracker sharedTracker] continueUserActivity:userActivity restorationHandler:restorationHandler];
  return YES;
}


@end

File: android/src/main/java/com/sarpongkb/flutterappsflyer/FlutterAppsflyerPlugin.java
package com.sarpongkb.flutterappsflyer;

import java.util.Map;
import java.util.HashMap;

import android.util.Log;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import com.appsflyer.AppsFlyerLib;
import com.appsflyer.AppsFlyerConversionListener;
import com.appsflyer.AppsFlyerProperties;
import com.appsflyer.CreateOneLinkHttpTask;
import com.appsflyer.share.LinkGenerator;
import com.appsflyer.share.ShareInviteHelper;

/** FlutterAppsflyerPlugin */
public class FlutterAppsflyerPlugin implements MethodCallHandler {
  private static final String DONE = "DONE";
  private static MethodChannel channel;
  private static Registrar registrar;

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    FlutterAppsflyerPlugin.registrar = registrar;
    channel = new MethodChannel(registrar.messenger(), "com.sarpongkb/flutter_appsflyer");
    channel.setMethodCallHandler(new FlutterAppsflyerPlugin());
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("initSdk")) {
      onInitSdk(call, result);
    } else if (call.method.equals("trackEvent")) {
      onTrackEvent(call, result);
    } else if (call.method.equals("getAppsflyerDeviceId")) {
      onGetAppsflyerDeviceId(call, result);
    } else if (call.method.equals("getSdkVersion")) {
      onGetSdkVersion(call, result);
    } else if (call.method.equals("generateShareLink")) {
      onGenerateShareLink(call, result);
    } else if (call.method.equals("setCustomerUserId")) {
      onSetCustomerUserId(call, result);
    } else if (call.method.equals("setUserEmails")) {
      onSetUserEmails(call, result);
    } else if (call.method.equals("setAdditionalData")) {
      onSetAdditionalData(call, result);
    } else if (call.method.equals("isDeviceTrackingEnabled")) {
      isDeviceTrackingEnabled(call, result);
    } else if (call.method.equals("setDeviceTrackingDisabled")) {
      onSetDeviceTrackingDisabled(call, result);
    } else {
      result.notImplemented();
    }
  }

  private void onInitSdk(MethodCall call, Result result) {
    String devKey = call.argument("devKey");
    String appleAppId = call.argument("appleAppId");
    boolean isDebug = call.argument("isDebug");
    AppsFlyerLib.getInstance().setDebugLog(isDebug);
    AppsFlyerLib.getInstance().init(devKey, conversionDataListener, registrar.context());
    AppsFlyerLib.getInstance().trackAppLaunch(registrar.context(), devKey);
    // AppsFlyerLib.getInstance().startTracking(registrar.activity());
    result.success(DONE);
  }

  private void onTrackEvent(MethodCall call, Result result) {
    String eventName = call.argument("eventName");
    Map eventValues = call.argument("eventValues");
    AppsFlyerLib.getInstance().trackEvent(registrar.context(), eventName, eventValues);
    result.success(DONE);
  }

  private void onGetAppsflyerDeviceId(MethodCall call, Result result) {
    String deviceId = AppsFlyerLib.getInstance().getAppsFlyerUID(registrar.context());
    result.success(deviceId);
  }

  private void onGetSdkVersion(MethodCall call, Result result) {
    String sdkVersion = AppsFlyerLib.getInstance().getSdkVersion();
    result.success(sdkVersion);
  }

  private void isDeviceTrackingEnabled(MethodCall call, Result result) {
    boolean disabled = AppsFlyerLib.getInstance().isTrackingStopped();
    result.success(!disabled);
  }

  private void onSetDeviceTrackingDisabled(MethodCall call, Result result) {
    boolean disabled = call.arguments();
    AppsFlyerLib.getInstance().setDeviceTrackingDisabled(disabled);
    result.success(DONE);
  }

  private void onGenerateShareLink(MethodCall call, final Result result) {
    String campaign = call.argument("campaign");
    String channel = call.argument("channel");
    String playerId = call.argument("playerId");
    String keywords = call.argument("keywords");

    LinkGenerator generator = ShareInviteHelper.generateInviteUrl(registrar.context());
    generator.setCampaign(campaign);
    generator.setChannel(channel);
    generator.setReferrerCustomerId(playerId);
    generator.addParameter("af_keywords", keywords);

    generator.generateLink(registrar.context(), new CreateOneLinkHttpTask.ResponseListener() {
      @Override
      public void onResponse(String linkUrl) {
        result.success(linkUrl);
      }

      @Override
      public void onResponseError(String errorMessage) {
        result.error("ONE_LINK_ERROR", errorMessage, null);
      }
    });
  }

  private void onSetCustomerUserId(MethodCall call, Result result) {
    String id = call.arguments();
    AppsFlyerLib.getInstance().setCustomerUserId(id);
    result.success(DONE);
  }

  private void onSetUserEmails(MethodCall call, Result result) {
    String userEmails = call.arguments().toString(); // need to convert array/list to a string
    AppsFlyerLib.getInstance().setUserEmails(AppsFlyerProperties.EmailsCryptType.NONE, userEmails);
    result.success(DONE);
  }

  private void onSetAdditionalData(MethodCall call, Result result) {
    HashMap<String, Object> data = call.arguments();
    AppsFlyerLib.getInstance().setAdditionalData(data);
    result.success(DONE);
  }

  private AppsFlyerConversionListener conversionDataListener = new AppsFlyerConversionListener() {
    public void onInstallConversionDataLoaded(Map<String, String> conversionData) {
      Log.d("FlutterAppsflyer onInstallConversionDataLoaded", conversionData.toString());
      channel.invokeMethod("conversionDataReceived", conversionData);
    }

    public void onAppOpenAttribution(Map<String, String> attributionData) {
      Log.d("FlutterAppsflyer onAppOpenAttribution", attributionData.toString());
      channel.invokeMethod("appOpenAttribution", attributionData);
    }

    public void onInstallConversionFailure(String errorMessage) {
      Log.d("FlutterAppsflyer onInstallConversionFailure", errorMessage);
      channel.invokeMethod("conversionDataRequestFailure", errorMessage);
    }

    public void onAttributionFailure(String errorMessage) {
      Log.d("FlutterAppsflyer onAttributionFailure", errorMessage);
      channel.invokeMethod("appOpenAttributionFailure", errorMessage);
    }
  };
}
